src/Symfony/Component/Validator/Constraints/HostnameValidator.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+/**
+ * @author Dmitrii Poddubnyi <dpoddubny@gmail.com>
+ */
+class HostnameValidator extends ConstraintValidator
+{
+    /**
+     * https://tools.ietf.org/html/rfc2606.
+     */
+    private const RESERVED_TLDS = [
+        'example',
+        'invalid',
+        'localhost',
+        'test',
+    ];
+
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof Hostname) {
+            throw new UnexpectedTypeException($constraint, Hostname::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $value = (string) $value;
+        if ('' === $value) {
+            return;
+        }
+        if (!$this->isValid($value) || ($constraint->requireTld && !$this->hasValidTld($value))) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Hostname::INVALID_HOSTNAME_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function isValid(string $domain): bool
+    {
+        return false !== filter_var($domain, FILTER_VALIDATE_DOMAIN, FILTER_FLAG_HOSTNAME);
+    }
+
+    private function hasValidTld(string $domain): bool
+    {
+        return false !== mb_strpos($domain, '.') && !\in_array(mb_substr($domain, mb_strrpos($domain, '.') + 1), self::RESERVED_TLDS, true);
```

Comment:
This should not use `mb_*` functions.

Reply:
Is there a guarantee that it's not a multi-byte string?
@lyrixx suggested to use `mb_`
![image](https://user-images.githubusercontent.com/1675033/68119443-683c0d80-ff0b-11e9-9b1e-4a20207d3bfb.png)

Reply:
We don't care about mb or not here, at all. Why would we?

Reply:
@nicolas-grekas  Got it, you are right - we only splitting the line here using the dot position.
![image](https://user-images.githubusercontent.com/1675033/68291667-04445100-0093-11ea-98d2-78907f83bc55.png)